src/spi2.rs

```diff
@@ -0,0 +1,754 @@
+                read_chunk.as_mut_ptr(),
+                write_chunk.as_ptr(),
+                max(read_chunk.len(), write_chunk.len()),
+                read_chunk.len(),
+            )?;
+        }
+
+        match read.len().cmp(&write.len()) {
+            cmp::Ordering::Equal => { /* Nothing left to do */ }
+            cmp::Ordering::Greater => {
+                // Read remainder
+                self.read(&mut read[write.len()..])?;
+            }
+            cmp::Ordering::Less => {
+                // Write remainder
+                self.write(&write[read.len()..])?;
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn transfer_in_place(&mut self, words: &mut [u8]) -> Result<(), EspError> {
+        for chunk in words.chunks_mut(self.trans_len) {
+            let ptr = chunk.as_mut_ptr();
+            let len = chunk.len();
+            self.polling_transmit(ptr, ptr, len, len)?;
+        }
+
+        Ok(())
+    }
+
+    pub fn flush(&mut self) -> Result<(), EspError> {
+        // Since we use polling transactions, flushing isn't required.
+        // In future, when DMA is available spi_device_get_trans_result
+        // will be called here.
+        Ok(())
+    }
+
+    fn polling_transmit(
+        &mut self,
+        read: *mut u8,
+        write: *const u8,
+        transaction_length: usize,
+        rx_length: usize,
+    ) -> Result<(), EspError> {
+        polling_transmit(
+            self.handle,
+            read,
+            write,
+            transaction_length,
+            rx_length,
+            self.hardware_cs,
+        )
+    }
+
+    /// Empty transaction to de-assert CS.
+    pub fn finish(&mut self) -> Result<(), EspError> {
+        polling_transmit(self.handle, ptr::null_mut(), ptr::null(), 0, 0, false)
+    }
+}
+
+impl<'d> embedded_hal::spi::ErrorType for SpiBusMasterDriver<'d> {
+    type Error = SpiError;
+}
+
+impl<'d> SpiBusFlush for SpiBusMasterDriver<'d> {
+    fn flush(&mut self) -> Result<(), Self::Error> {
+        SpiBusMasterDriver::flush(self).map_err(to_spi_err)
+    }
+}
+
+impl<'d> SpiBusRead for SpiBusMasterDriver<'d> {
+    fn read(&mut self, words: &mut [u8]) -> Result<(), Self::Error> {
+        SpiBusMasterDriver::read(self, words).map_err(to_spi_err)
+    }
+}
+
+impl<'d> SpiBusWrite for SpiBusMasterDriver<'d> {
+    fn write(&mut self, words: &[u8]) -> Result<(), Self::Error> {
+        SpiBusMasterDriver::write(self, words).map_err(to_spi_err)
+    }
+}
+
+impl<'d> SpiBus for SpiBusMasterDriver<'d> {
+    fn transfer(&mut self, read: &mut [u8], write: &[u8]) -> Result<(), Self::Error> {
+        SpiBusMasterDriver::transfer(self, read, write).map_err(to_spi_err)
+    }
+
+    fn transfer_in_place(&mut self, words: &mut [u8]) -> Result<(), Self::Error> {
+        SpiBusMasterDriver::transfer_in_place(self, words).map_err(to_spi_err)
+    }
+}
+
+pub struct SpiMasterDriver<'d> {
+    pub(crate) handle: spi_host_device_t,
+    pub(crate) lock: CriticalSection,
```

Comment:
Should this be moved to `SpiConfigPool`?

Reply:
wdym by move? moving the lock to the ConfigPool. Moving the complete struct to the ConfigPool?

Reply:
Yes just the lock. It's only needed by the ConfigPool.

Reply:
if you want to use the driver on multiple threads we still need a looking mechanism for the default EspSpiDevice. and the EspSpiDevice does't now anything about a ConfigPool

Reply:
Locking isn't necessary, the spi api is thread safe.

Reply:
it would be nice if we could ask Borrow<EspSpiDevice> if it gives us a T &T or &mut T than we know we dont need an lock, only need a lock if we got an Arc(T)

Reply:
Regardless of the type of `Borrow`, you don't need to lock. Even if you have devices on different threads. esp-idf's SPI api is thread safe, so we're free to share the bus without locking around it.
Each individual device can't be shared between threads but each device can be used on a separate thread from other devices.

Reply:
where do you take that it is thread safe ? it states in big read letters that it is not in the documentation? 
From the documentation:
```
The SPI master driver has the concept of multiple Devices connected to a single bus (sharing a single ESP32 SPI peripheral). As long as each Device is accessed by only one task, the driver is thread safe. However, if multiple tasks try to access the same SPI Device, the driver is not thread-safe. In this case, it is recommended to either:

    Refactor your application so that each SPI peripheral is only accessed by a single task at a time.

    Add a mutex lock around the shared Device using [xSemaphoreCreateMutex](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/freertos.html#c.xSemaphoreCreateMutex).

```

Reply:
is it because we currently need a &mut T to EspSpiDevice for transaction and we have no current way of duplicating it / sending it to another thread ?

Reply:
> where do you take that it is thread safe ? it states in big read letters that it is not in the documentation?

I'm confused. That snippet of documentation you've copied says it's thread safe...
The bus can be shared but devices cannot.